### ISSUE TITLE
Notification silent flag for web notifications

### ISSUE THREAD

**Issue by collimarco** [mozilla-mobile/fenix#18145]
[comment]: # (Please do your best to search for duplicate issues before filing a new issue so we can keep our issue board clean)
[comment]: # (Every issue should have exactly one feature request described in it. Please do not file feedback list tickets as it is difficult to parse them and address their individual points)
[comment]: # (Feature Requests are better when they’re open-ended instead of demanding a specific solution e.g: “I want an easier way to do X” instead of “add Y”)
[comment]: # (Read https://github.com/mozilla-mobile/fenix#i-want-to-file-an-issue for more information)

### What is the user problem or growth opportunity you want to see solved?

As described in the Web Notifications API standard, the browser should support a silent flag for notifications:

```
serviceWorkerRegistration.showNotification('Example', { silent: true })
```

Currently Firefox ignores this flag and always plays a sound.

### How do you know that this problem exists today? Why is this important?

At Pushpad we received requests from customers for this option. Chrome and derivates already support it and should be easy to implement in Firefox.

### Who will benefit from it?

Users will benefit from it since they can receive updates that are not time-sensitive, without disturbing them for each notification.

Please fix this since it is important for the user experience: e.g. a web app may want to mute a conversation, or a news website prefers to disable the sound since notifications are not urgent, etc.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-9801)

**Comment by codrut-topliceanu** [comment]
I've looked into this a bit and it seems that on Chrome the silent flag is ignored and it just uses the app notification settings for whether or not to add sound to the notification. They do however seem to support the flag if [this site is correct about it](https://web-push-book.gauntface.com/demos/notification-examples/) (button is enabled). I'm not entirely sure how the site knows if the browser supports the flag or not, if there's a communication going on or if it just checks the browser against a list.

1. We can, of course, look in the notification tags and see if `silent = true` is present and override the app settings accordingly, but are we sure we'd want that? (UX question)
2. To add support the flag I think is a question for GeckoView. (https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/showNotification)

**Comment by collimarco** [comment]
Thanks for looking into this issue.

>  it seems that on Chrome the silent flag is ignored

Chrome and other Chromium-based browsers **use the silent flag properly** on desktop. 

On mobile they are going to implement it:
https://bugs.chromium.org/p/chromium/issues/detail?id=1181685

Consider that the silent flag is already part of the standard:
https://notifications.spec.whatwg.org/#silent-preference-flag

And it can be useful to improve the user experience in many cases:
https://blog.pushpad.xyz/2021/03/pushpad-the-new-feature-for-sending-silent-notifications/

> we sure we'd want that? (UX question)

Yes, also because a web app may want to send some notifications with sound (e.g. important or direct messages), while muting others (e.g. group conversations that the user has muted - like Whatsapp).

**Comment by Mugurell** [comment]
>>we sure we'd want that? (UX question)

> Yes, also because a web app may want to send some notifications with sound (e.g. important or direct messages), while muting others (e.g. group conversations that the user has muted - like Whatsapp).

About this,
Android already allows and Fenix offers the posibility of users setting if all incoming website notifications should be silent or not with a simple toggle in system settings.
Chrome goes even further and allows setting a different behavior for each website.

And then this new flag would come to control the same thing.
So I guess the question for UX is how should these two be intersected? Which should have priority?

I'd propose the following behavior:

| app permits audible notifications | website sends audible notification| notification may ring or vibrate the device depending on phone setting |
| :-: | :-: | :-: |
| ✓ | ✓ | ✓ |
| ✓ | ✗ | ✗ |
| ✗ | ✓ | ✗ |
| ✗ | ✗ | ✗ |

**Comment by collimarco** [comment]
Yes, when I read the standard I expected the behavior described in your table: if the website OR the user decides to mute the notifications, then mute the notifications.

**Comment by codrut-topliceanu** [comment]
Fixed in GV. Waiting on merge.

**Comment by amedyne** [comment]
@codrut-topliceanu Can you check if this has been fixed?

**Comment by codrut-topliceanu** [comment]
`GeckoViewWebNotification` `silent` flag has been introduced 🥳 

![image](https://user-images.githubusercontent.com/60002907/123246610-a5914c80-d4ee-11eb-8b9a-862c29134c3c.png)

Now we should be able to use it in AC in: 
`WebNotification`, `GeckoWebNotificationDelegate`, `NativeNotificationBridge` and `WebNotificationFeature`.
There we can set the importance of the notification. At the moment we use `IMPORTANCE_LOW` for web notifications.

![image](https://user-images.githubusercontent.com/60002907/123249417-99f35500-d4f1-11eb-8319-3fa9b70ec620.png)

With approval from UX we can change that so that web notifications' priority depends on the silent flag. 
Pinging @rocketsroger as he may have valuable input on this.

**Comment by rocketsroger** [comment]
One possible solution is to change default to `IMPORTANCE_DEFAULT` but if silent flag is true then set it to `IMPORTANCE_LOW`.

**Comment by Mugurell** [comment]
This would still be a nice to have.

**Comment by csadilek** [comment]
Moved to bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1809284

Change performed by the [Move to Bugzilla add-on](https://addons.mozilla.org/en-US/firefox/addon/move-to-bugzilla/).

**Comment by Verdi** [comment]
Commenting for UX - this looks good.
